### ISSUE TITLE
$PerlIO::encoding::fallback = FB_DEFAULT leads to duplicated output

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7309]
Migrated from [rt.perl.org#29720](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=29720) (status was 'open')

Searchable as RT29720$

**Comment by p5pRT** [comment]
### From aa29@mail.ru

To&#8203;: perlbug@&#8203;perl\.org
Subject&#8203;: $PerlIO&#8203;::encoding&#8203;::fallback = FB\_DEFAULT leads to duplicated output
Reply\-To&#8203;: aa29@&#8203;mail\.ru
Message\-Id&#8203;: \<5\.8\.4\_2160\_1084980780@&#8203;INFORMED>

This is a bug report for perl from aa29@&#8203;mail\.ru\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.4\.

$PerlIO&#8203;::encoding&#8203;::fallback = FB\_DEFAULT leads to duplicated output\.

It is possible to change check\-mode via $PerlIO&#8203;::encoding&#8203;::fallback&#8203;:

use Encode qw\(&#8203;:fallback\_all\);
use encoding 'utf8';

$PerlIO&#8203;::encoding&#8203;::fallback = FB\_DEFAULT;

binmode\(STDERR\, "&#8203;:encoding\(cp866\)"\);
warn "foobar";

This code gives four messages instead of one&#8203;:

foobar at 6\.pl line 7\.
foobar at 6\.pl line 7\.
foobar at 6\.pl line 7\.
foobar at 6\.pl line 7\.

And with redirection STDERR to file it gives three messages&#8203;:

foobar at 6\.pl line 7\.
foobar at 6\.pl line 7\.
foobar at 6\.pl line 7\.

Further investigation shows that there is no duplication if
$PerlIO&#8203;::encoding&#8203;::fallback = FB\_DEFAULT | FB\_PERLQQ; \# or FB\_\(HT|X\)MLCREF

Looking into ext\\Encode\\Encode\.xs I found such code
\(ext\\Encode\\Encode\.xs\, line 229\)&#8203;:

&nbsp;    if \(check && \!\(check & ENCODE\_LEAVE\_SRC\)\)\{
 sdone = SvCUR\(src\) \- \(slen\+sdone\);
 if \(sdone\) \{
&nbsp;     sv\_setpvn\(src\, \(char\*\)s\+slen\, sdone\);
 \}
 SvCUR\_set\(src\, sdone\);
&nbsp;    \}

If check is set to FB\_DEFAULT \(which is 0\) and no other fallback is
defined\, then it behaves as if ENCODE\_LEAVE\_SRC is set\, and buffer
does not became truncated\, and then it will be flushed several times\.


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl v5\.8\.4&#8203;:

Configured by aa29 at Mon May 17 17&#8203;:59&#8203;:46 2004\.

Summary of my perl5 \(revision 5 version 8 subversion 4\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=4\.0\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    usethreads=undef use5005threads=undef useithreads=define
usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, ccflags
='\-nologo \-Gf \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32 \-D\_CONSOLE \-DNO\_STRICT \-DHAVE
\_DES\_FCRYPT  \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO \-DPERL
\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion=''\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='\_\_int64'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags
'\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf  \-libpath&#8203;:"c&#8203;:\\perl\\lib\\CORE"  \-ma
chine&#8203;:x86'
&nbsp;    libpth=\\lib
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib
uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib
msvcrt\.lib
&nbsp;    perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib
uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib
msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=yes\, libperl=perl58\.lib
&nbsp;    gnulibc\_version='undef'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\,
lddlflags='\-dll \-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf  \-libpath&#8203;:"c&#8203;:\\perl
\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:




---
@&#8203;INC for perl v5\.8\.4&#8203;:
&nbsp;    C&#8203;:/Perl/lib
&nbsp;    C&#8203;:/Perl/site/lib
&nbsp;    \.



---
Environment for perl v5\.8\.4&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)

PATH=C&#8203;:\\cygwin\\bin;C&#8203;:\\Tcl\\bin;C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\Syst
em32\\Wbem;C&#8203;:\\Perl\\bin;C&#8203;:\\Program Files\\Support Tools;D&#8203;:\\src\\lib;C&#8203;:\\Program
Files\\Microsoft Visual Studio\\Common\\Tools\\WinNT;C&#8203;:\\Program Files\\Microsoft
Visual Studio\\Common\\MSDev98\\Bin;C&#8203;:\\Program Files\\Microsoft Visual
Studio\\Common\\Tools;C&#8203;:\\Program Files\\Microsoft Visual
Studio\\VC98\\bin;C&#8203;:\\Arc;C&#8203;:\\Program Files\\Utils;C&#8203;:\\Mysql\\bin;C&#8203;:\\Program
Files\\Debugging Tools for Windows;C&#8203;:\\Tcl\\bin;D&#8203;:\\Linda\\XML\\fop;C&#8203;:\\Program
Files\\GNU\\WinCvs 1\.2;D&#8203;:\\src\\bin;C&#8203;:\\Program Files\\Far
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

aa29

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

On Wed May 19 08&#8203;:34&#8203;:56 2004\, aa29 wrote&#8203;:
> $PerlIO&#8203;::encoding&#8203;::fallback = FB\_DEFAULT leads to duplicated output\.

This old bug still exists in bleadperl (as of a few days ago). It also
exists in all of perl 5.14.2, 5.12.3, 5.10.1. I have tested on
amd64-linux with this command.

```
$ ~/local/perlblead/bin/perl5.15.4 -we 'use Encode; use
PerlIO::encoding; $PerlIO::encoding::fallback = Encode::FB_XMLCREF();
binmode STDOUT, "encoding(iso-8859-2)" or die; print
"\x{e9}l\x{151}.u\x{ef} \x{2203}t\n";'
élő.u&#xef; &#x2203;t
élő.u&#xef; &#x2203;t
élő.u&#xef; &#x2203;t
$
```

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

Besides printing duplicate output\, a filehandle with an encoding layer
with fallback set also usually raises an exception "Close with partial
character" when you try to close it\.  This error message is not
documented in either perldiag or PerlIO&#8203;::encoding\, and\, in any case\,
there shouldn't be an error\.

I attach a test script that tests whether this bug is still present&#8203;:
it tests for both correct output and no exception when you close the
file\.

Ambrus

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

[unencodable.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1026226/506146/unencodable.pl)

**Comment by toddr** [comment]
@Leont does that mean you're looking at this issue then?

**Comment by Leont** [comment]
I have some ideas, but it may require some work on the `Encode` side too; `FB_DEFAULT` having a double meaning is inconvenient.

**Comment by toddr** [comment]
OK. I’m gonna put your name on it so we know who is involved

**Comment by osir3z** [comment]
After trying `:encoding(...)` and finding the warnings and `perlqq` output unsatisfactory, I discoved `$PerlIO::encoding::fallback` (which btw should be easier to find out about) and tried setting it to `FB_DEFAULT`. Of course, that's when I discovered the issue(s) discussed here.


## Workaround

After several hours investigating and experimenting, I've come up with a workaround:
> When setting `$PerlIO::encoding::fallback` always ensure that the `LEAVE_SRC` bit is clear and the `STOP_ON_PARTIAL` bit is set.

For example, to set `$PerlIO::encoding::fallback` to `$fallback` use:
```perl
$PerlIO::encoding::fallback = (($fallback) & ~Encode::LEAVE_SRC()) | Encode::STOP_ON_PARTIAL();
```
_(unnecessary parentheses included for clarity)_


## Justification

Note that you'll find this same construct in [`PerlIO-encoding/encoding.xs#L175`](https://metacpan.org/release/RJBS/perl-5.36.0/source/ext/PerlIO-encoding/encoding.xs#L175). And whilst it affects the initial value of `$PerlIO::encoding::fallback` set by the module, it clearly doesn't affect it every time the value is modified by the user.

In my experimentation I initially found that clearing the `LEAVE_SRC` bit and setting an "unused" bit (e.g. `0x8000`), to ensure the `LEAVE_SRC` bit was honored (see [Encode#LEAVE_SRC](https://perldoc.perl.org/Encode#LEAVE_SRC)), solved the issue. But after seeing the code in `PerlIO-encoding/encoding.xs` discussed above, I think it's safe to suggest that the `STOP_ON_PARTIAL` bit should be set instead of an "unused" bit.


## Testing

I modified a copy of the `unencodable.pl` test suite kindly provided by ambrus@math.bme.hu (see https://github.com/Perl/perl5/issues/7309#issuecomment-543944094 above). The only place where `$PerlIO::encoding::fallback` is set in that code is line 78, and by modifying it from:
```perl
if (!$nofb) { $PerlIO::encoding::fallback = $fb; }
```
to:
```perl
if (!$nofb) { $PerlIO::encoding::fallback = (($fb) & ~Encode::LEAVE_SRC()) | Encode::STOP_AT_PARTIAL(); }
```
all the tests now pass.

_(tested with Perl v5.30.2 on Windows 10, Perl v5.30.3 on Ubuntu 20.04 LTS for WSL2, and Perl v5.28.1 on Debian Buster)_

**Comment by osir3z** [comment]
Thanks to @Leont this issue should be resolved in Perl 5.34.0 and later.

The solution allows you to set whatever value you like for `$PerlIO::encoding::fallback`, but every time you use `:encoding(...)`, that value is sanitised (using the same logic as the workaround below) before it is actually used by the encoder/decoder.


## Workaround

For versions before Perl 5.34.0, always clear the `LEAVE_SRC` bit and set the `STOP_AT_PARTIAL` bit when setting `$PerlIO::encoding::fallback`, e.g.:
```perl
$PerlIO::encoding::fallback = (($fallback) & ~Encode::LEAVE_SRC()) | Encode::STOP_AT_PARTIAL();
```
_(tested with Perl 5.30.2 on Windows 10, Perl 5.30.3 on Ubuntu 20.04 LTS for WSL2, and Perl 5.28.1 on Debian Buster)_


## Background

When I encountered this issue a couple of days ago, I was trying to set `$PerlIO::encoding::fallback` to `FB_DEFAULT` because I was unhappy with the `qq`-style output and the warnings I got when I used `:encoding(...)`. Obviously, as per this issue, that resulted in duplicated output.

After some experimentation I discovered that clearing the `LEAVE_SRC` bit resolved the duplicated output for all but `FB_DEFAULT`. But that's because `LEAVE_SRC` is only honored when `$PerlIO::encoding::fallback` is set (see [Encode#LEAVE_SRC](https://perldoc.perl.org/Encode#LEAVE_SRC)). Testing showed that by forcing an "unused" bit (e.g. `0x8000`) to be set, the clear `LEAVE_SRC` bit would be honored and everything appeared to work.

Unhappy at having to hack a solution with an "unused" bit that may someday get used, I dug in to the code for `PerlIO::encoding` on MetaCPAN and found @Leont's code which sanitised `$PerlIO::encoding::fallback` according to the logic in the workaround above (see [`PerlIO-encoding/encoding.xs#L175`](https://metacpan.org/release/XSAWYERX/perl-5.34.0/source/ext/PerlIO-encoding/encoding.xs#L175)). I assumed that it wasn't working for some reason, but it turns out that it was just the latest version of the code which wasn't included in the versions of Perl that I was testing on.

Looking at various different version of Perl going back through the years, it is clear that the default value for `$PerlIO::encoding::fallback` always has a clear `LEAVE_SRC` bit and a set `STOP_AT_PARTIAL` bit. Obviously @Leont came to the same conclusion. Thankfully, using this combination means I avoid using a hack, and also likely avoid some errors I hadn't yet encountered.

**Comment by jkeenan** [comment]
> Thanks to @Leont this issue should be resolved in Perl 5.34.0 and later.
> 

@Leont, do you concur?

**Comment by Leont** [comment]
> @Leont, do you concur?

Yeah, this is solved.
